tools/common/CoreResolver.cs

```diff
@@ -0,0 +1,121 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+#if MTOUCH
+using ProductException=Xamarin.Bundler.MonoTouchException;
+#else
+using ProductException=Xamarin.Bundler.MonoMacException;
+#endif
+
+namespace Xamarin.Bundler {
+
+	public abstract class CoreResolver : IAssemblyResolver {
+
+		internal Dictionary<string, AssemblyDefinition> cache;
+
+		public string FrameworkDirectory { get; set; }
+		public string RootDirectory { get; set; }
+		public string ArchDirectory { get; set; }
+
+
+		public CoreResolver ()
+		{
+			cache = new Dictionary<string, AssemblyDefinition> (NormalizedStringComparer.OrdinalIgnoreCase);
+		}
+
+		public IDictionary<string, AssemblyDefinition> ResolverCache { get { return cache; } }
+
+		public IDictionary ToResolverCache ()
+		{
+			var resolver_cache = new Dictionary<string, AssemblyDefinition> (NormalizedStringComparer.OrdinalIgnoreCase);
+			foreach (var pair in cache)
+				resolver_cache.Add (pair.Key, pair.Value);
+
+			return resolver_cache;
+		}
+
+		public void Dispose ()
+		{
+		}
+
+		public AssemblyDefinition Resolve (AssemblyNameReference name)
+		{
+			return Resolve (name, new ReaderParameters { AssemblyResolver = this });
+		}
+
+		public abstract AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters parameters);
+
+		protected ReaderParameters CreateDefaultReaderParameters (string path)
+		{
+			var parameters = new ReaderParameters ();
+			parameters.AssemblyResolver = this;
+			parameters.InMemory = new FileInfo (path).Length < 1024 * 1024 * 100; // 100 MB.
+			parameters.ReadSymbols = true;
+			parameters.SymbolReaderProvider = new DefaultSymbolReaderProvider (throwIfNoSymbol: false);
+			return parameters;
+		}
+
+		public virtual AssemblyDefinition Load (string fileName)
+		{
+			if (!File.Exists (fileName))
+				return null;
+
+			AssemblyDefinition assembly;
+			var name = Path.GetFileNameWithoutExtension (fileName);
+			if (cache.TryGetValue (name, out assembly))
+				return assembly;
+
+			try {
+				fileName = Target.GetRealPath (fileName);
+
+				// Check the architecture-specific directory
+				if (Path.GetDirectoryName (fileName) == FrameworkDirectory && !string.IsNullOrEmpty (ArchDirectory)) {
+					var archName = Path.Combine (ArchDirectory, Path.GetFileName (fileName));
+					if (File.Exists (archName))
+						fileName = archName;
+				}
+
+				var parameters = CreateDefaultReaderParameters (fileName);
+				try {
+					assembly = ModuleDefinition.ReadModule (fileName, parameters).Assembly;
+				}
+				catch (InvalidOperationException e) {
+					// cecil use the default message so it's not very helpful to detect the root cause
+					if (!e.TargetSite.ToString ().Contains ("Void ReadSymbols(Mono.Cecil.Cil.ISymbolReader)"))
+						throw;
+					ErrorHelper.Show (ErrorHelper.CreateWarning (129, $"Debugging symbol file for '{fileName}' does not match the assembly and is ignored."));
```

Comment:
If you show the warning after (successfully) loading the assembly without debug symbols, you won't end up showing the warning if there's something else wrong with the assembly (in which case it would be confusing).

Reply:
good point, telling the symbols are bad and then failing to load the assembly is confusing
I'll re-order this